core/pfe-core/controllers/tabs-aria-controller.ts

```diff
@@ -0,0 +1,123 @@
+import type { ReactiveController, ReactiveControllerHost } from 'lit';
+
+import { Logger } from '@patternfly/pfe-core/controllers/logger.js';
+
+export interface TabsControllerOptions<Tab, Panel> {
+  /** Add an `isTab` predicate to ensure this tabs instance' state does not leak into parent tabs' state */
+  isTab: (node: unknown) => node is Tab;
+  isActiveTab: (tab: Tab) => boolean;
+  /** Add an `isPanel` predicate to ensure this tabs instance' state does not leak into parent tabs' state */
+  isPanel: (node: unknown) => node is Panel;
+  getHTMLElement?: () => HTMLElement;
+}
+
+export class TabsAriaController<
+  Tab extends HTMLElement = HTMLElement,
+  Panel extends HTMLElement = HTMLElement,
+> implements ReactiveController {
+  #logger: Logger;
+
+  #host: ReactiveControllerHost;
+
+  #element: HTMLElement;
+
+  #tabPanelMap = new Map<Tab, Panel>();
+
+  #options: TabsControllerOptions<Tab, Panel>;
+
+  #mo = new MutationObserver(this.#onSlotchange.bind(this));
+
+  get tabs() {
+    return [...this.#tabPanelMap.keys()] as Tab[];
+  }
+
+  get activeTab(): Tab | undefined {
+    return this.tabs.find(x => this.#options.isActiveTab(x));
+  }
+
+  /**
+   * @example Usage in PfTab
+   *          ```ts
+   *          new TabsController(this, {
+   *             isTab: (x): x is PfTab => x instanceof PfTab,
+   *             isPanel: (x): x is PfTabPanel => x instanceof PfTabPanel
+   *          });
+   *          ```
+   */
+  constructor(
+    host: ReactiveControllerHost,
+    options: TabsControllerOptions<Tab, Panel>,
+  ) {
+    this.#options = options;
+    this.#logger = new Logger(host);
+    if (host instanceof HTMLElement) {
+      this.#element = host;
+    } else {
+      const element = options.getHTMLElement?.();
+      if (!element) {
+        throw new Error('TabsController must be instantiated with an HTMLElement or a `getHTMLElement()` option');
+      }
+      this.#element = element;
+    }
+    (this.#host = host).addController(this);
+    this.#element.addEventListener('slotchange', this.#onSlotchange);
```

Comment:
Can you explain how this option of `getHTMLElement` would be different in implementation from `host` given we are attaching a `slotchange` event to the element?  IIRC `slotchange` would only work given the element is a custom element with a shadowroot so wouldn't that also just be a host?  Is this a way to instantiate this in a parent element outside of that target?

```
<outer-parent> (has tabs-aria-controller setup with getHTMLElement targetting inner-tabs)
  <inner-tabs>
     <im-a-tab></im-a-tab>
     <im-a-panel></im-a-panel>
     ...
  <inner-tabs>
</outer-parent>
```

Reply:
Here's an example from haunted, a react-a-like ce framework

https://github.com/matthewp/haunted/pull/239/files#diff-c5df8cacedd70d19bd53b47ac1cf3de080ba8b014f4f8c9ba73f7b70fecdb305

Similarly for full-menlo react: https://lit.dev/docs/frameworks/react/#usecontroller
And https://github.com/lit/lit/blob/main/packages/react/src/use-controller.ts

TL;DR: reactive controller hosts don't have to be custom elements, they just have to be POJOs which implement a particular interface. If we can avoid requiring our controllers to be hosted by elements, our controllers will be of use in more places

Reply:
So why slot change?

Maybe the controller host object (which owns the update lifecycle) isn't the same as the rendering custom element

Maybe they're are Very Important Reasons for someone to reimplement our ces (like... It's embedded in their dashboard rendering framework), and they use their own unbranded ces, which aren't reactive elements, but maybe we could create host objects which play puppeteer with the real ces.